### PR TITLE
Jkarni/content types

### PR DESCRIPTION
**Pull request by jkarni**
Do not merge yet.

Support for multiple content-types, varying based on the Accept param ( #9 ).

### CODE REVIEW

**Comment by mf59816** [comment]
@alpmestan: (re: last comment) I guess if in doubt, keep it simple, so leave it as it is nows?

I think I may want to use this soon.  What can I do something to help getting it into a new release?  Is there anything missing besides the merge conflict resolution and possibly a tighter version constraint for the servant package?

**Comment by jkarni** [comment]
What's missing for a release is, at a minimum:
(1) some resolution to QQ with the addition of content-types (and headers) in `servant`. I would like to do this before even merging this. Don't work on this without talking to me first, since there are some design decisions we should consider.
(2) update servant-client, servant-jquery and servant-docs updates to go along with the content-types additions (in both reqbody and response). This shouldn't be too hard, and is up for grabs, at least till this weekend.

**Comment by alpmestan** [comment]
I've given this some more thoughts. Maybe we should just take the path of least pain here, and have all the multiple-content-types-enabled combinators end with a prime (`'`), and have the JSON-only ones be named as they are now? This is, as far as I can tell, what will break the smallest amount of code, and probably none at all in some cases (although some code might break with all these changes that are coming in).

tl;dr:

``` haskell
-- multiple content types support
data Delete' (cts :: [*]) a
data Get' (cts :: [*]) a
data Post' (cts :: [*]) a
data Put' (cts :: [*]) a

-- json-only
type Delete a = Delete' [JSON] a
type Get a = Get' [JSON] a
type Post a = Post' [JSON] a
type Put a = Put [JSON] a

-- and same for ReqBody
```

What's your opinion on this guys? This would be in the "make the simplest scenarios simple to write" spirit.

**Comment by jkarni** [comment]
I was thinking that we could name the ones with explicit content-type with a single quote, but then *not* provide any type alias. If a user wants to then be (almost) backward compatible, they could add it themselves. That way, there's an open area in the namespace for every type synonym, whether JSON, HTML, or whatever.

**Comment by mf59816** [comment]
I think `[JSON]` may not be the simplest scenario for the majority.  @jkarni's approach is a bit unusual, but I like that it sort of leaves a hole in the library that can be filled out by the user as convenient.

The haddock comments of `Get'` and friends should give code examples and explain that you're supposed to write our own `Get` for what you as servant user need most often.

**Comment by alpmestan** [comment]
I do think JSON is the most common for most of our users so far. Especially considering that's the only one we provide :p

But if you really don't like the idea, we'll do without it :) It's just that everyone using servant so far will go through the same process of defining their own simple Get, Post, etc, to keep their current code valid for the new version.

**Comment by soenkehahn** [comment]
I don't really like the idea to leave a hole in the library (as @mf59816 put it aptly). I think we should provide ready-made types for writing a JSON-only api.

Apart from that I don't really like the `'` naming convention as it doesn't really convey more information than that something is a variant of something else.

What about `type Get a = GetMulti [JSON] a`?

**Comment by jkarni** [comment]
I like `GetMulti`, but it's a bit long. I was originally in favour of just `Get` for the multiple CTs version.

My primary problem with defaulting to JSON-only isn't just that we give priority to JSON over, say HTML. (Although that's a big part of it -- while we ourselves use servant for webservices, I don't see why we should relegate HTML websites to the back of our minds when designing the library.) It's that we also give priority to JSON over, say, JSON *and* XML. And honestly I think an advantage of this entire infrastructure is that you can get XML (or YAML even -- does anyone do that?) for free, so you might as well, and then you want another type synonym.

**Comment by alpmestan** [comment]
@jkarni Think of it as a trivial migration path from servant 0.2 then, maybe? If we make `Get` the JSON-only version, no user has to change his/her code. And I liked `Get'` because it's just one more character to type to unleash the power of content types :) But if you guys have any better suggestion I'm all ears!

**Comment by mf59816** [comment]
What about `Get` instead of `GetMulti` and `GetJ` instead of the curren `Get`?  That would also suggest the convenient `GetX`, `GetJX` for people who need them.

**Comment by mf59816** [comment]
(Personally, I would value a sound design higher than upgrade convenience.  At least at this early stage.)

**Comment by alpmestan** [comment]
Well, here my final words on this:
- If we don't make `Get` be the JSON-only version, this will really break *everyone*'s code, and everyone will have to fix it *in the very same way*. This will break the examples from the repo, the docs, the getting started guide, the talk. All of them, just because of a naming convention.
- Right now, the few people who serve HTML serve files that communicate with the endpoints through javascript (and WIP through purescript). JSON is really the major user of `Get`, `Post` etc endpoints and will remain so for a while, I think. From this point of view (which you may not agree with and I can understand that), the simplest usage would benefit from being the simplest to write/type.
- My idea of using `Get'` is simply about giving access to richer content types by adding a single char. This is completely trivial to explain in the docs. "You want to serve this json but other formats too? Use `Get'` instead of `Get` and specify the formats you want with a list, the first format of the list being the default when no `Accept` header is given."

I think all the arguments have been articulated...? Or maybe @christian-marie and the rest of the Anchor team would have something to say that we've not thought of?

I guess the solution is just to put the propositions together in a clear and understandable format and then vote (we the core devs + our users)?

**Comment by tvh** [comment]
I would really like this to land in some form. I looked at reimplementing our OAuth2 Server using servant, but gut stuck because there is no way to use `application/x-www-form-urlencoded` with Servant currently. I'd be happy to add support for that in requests once the shape of this API change is clear.

I think the default should be to specify exactly what you want. I'm not convinced we need we need shortcuts here, but if there should be, then I preferer @mf59816's suggestion.

This will break a lot of existing code for sure, but will lead to a clearer API in my opinion.

**Comment by jkarni** [comment]
@tvh mostly this is blocked on waiting `servant-client` to support multiple content-types, and `servant` to have extensible quasi-quoting (this change would break the QQ anyhow, and it's become increasingly clear that the QQ we currently have doesn't match the extensibility of the combinators, making it close to useless for larger projects, I guess). I've done quite a bit on the second front, but haven't started on `servant-client` yet, though it shouldn't be very hard.

**Comment by tvh** [comment]
I think the `Servant.Server.ContentTypes` module should live in `servant` instead. When building a client, most of that is needed as well.

**Comment by mf59816** [comment]
+1 on the last comment.

I am hoping to write a PR for `servant-client` later today or tomorrow.  If that interferes with anybody's current activities please yell.

@jkarni: I haven't looked at QQ yet, but it sounds like you are thinking about discontinuing it altogether?  (I am not planning to use QQ, I just want the rest of this to merge.  If that requires fixing QQ, I can try to help with that, too.)

**Comment by jkarni** [comment]
I'm thinking of making it a separate package - I guess that means we could release servant without it being ready, though it'd be nicer to have a cleaner upgrade path for those who are using it.

I'm open to the idea of moving some of this stuff into `servant`, though maybe not the render logic. @mf59816 @tvh if you have more info on what gets used in the client, I'd love to hear it. My initial intuition was that only the content type type itself would be shared, but I can already see how e.g. the `Accept` instances/class would be too.

**Comment by tvh** [comment]
I would say that all the `MimeRender` stuff is necessary as well. I think the client side is largely symmetrical since you have both support for `ReqBody` and for response. The only difference that I see is, that a server has to support all the content types, while a client only has to support one of them. So maybe the `AllCTRender`, `AllMimeRender`, ... can live in `servant-server`. I wouldn't mind having them in `servant` as well though, as those instances are really only combinators on the other typeclasses and not intimately related to the server.

**Comment by mf59816** [comment]
+1

I started working on this, and got stuck (I don't think it's a hard problem I'm stuck with, it's just late): 

https://github.com/zerobuzz/servant-client/tree/content-types

Note:
* doesn't compile yet.
* i made the package depend on servant-server in order to get on with it.  I want to remove that dependency later.
* the type-level element check should probably go elsewhere, right?  where?
* did I restrict things to be always of the same content type in the client now?  is there a way to avoid that?
* should I add some non-json-only tests?

**Comment by tvh** [comment]
Restricting the client to only use one content-type is definitely sub-optimal. I think the simplest solution would be to always pick the first element in the list and work with that. That way you could avoid crazy type-level magic and have something running quickly that definitely works for every API.

A solution to fix this could be something like:
```haskell
class BodyCTType a where
    type BodyCTTypeDecision a

class ResponseCTType a where
    type ResponseCTTypeDecision a
````

The names are maybe not ideal, but it should work. It would break though if the server decides to use different content types for the same thing (which I think should be an error anyway).

**Comment by mf59816** [comment]
I feel you have a point, but I'm not sure I understand.

Doesn't "pick the first element" imply that the client functions will be constrained to that first content type?

With my approach, you can construct as many client functions as you wish by calling client more than once with different proxies.  But of course you are aware of that.

I will take another look after breakfast.

**Comment by mf59816** [comment]
Ok, got it.  Sorry for being a bit thick.  To rephrase what you said:

```haskell
class ReqCT ctyps where
    type ReqCTChoice ctyps

class ClientCT ctyps where
    type ClientCTChoice ctyps

instance (ClientCT ctyps, MimeUnrender (ClientCTChoice ctyps) result)
      => HasClient (Get ctyps result) where
  type Client (Get ctyps result) = BaseUrl -> EitherT String IO result
```

But won't that produce a lot of orphan instances?  The caller to servant-client will make the choice which set of supported content types maps to which actual choice, and they will do that for all end-points in all apis that import their module.  Not sure this is what we want either...

**Comment by mf59816** [comment]
What about this?:

```haskell
class ReqCT (layout, ctyps) where
    type ReqCTChoice (layout, ctyps)

class RespCT (layout, ctyps) where
    type RespCTChoice (layout, ctyps)
```

**Comment by mf59816** [comment]
I tried again: see above.  Does anybody have yet a better idea?

**Comment by tvh** [comment]
That design would work, yes. One other option would be to provide 2 functions. 
`client` as before, which would always take the first contenttype of the list, and `clientWith`, which would take a ´Proxy` of the mapping type -> content-type as well. I don't have a solution for how to not end up with 2 completely seperate functions though...

**Comment by tvh** [comment]
@mf59816 have you made more progress on the client side? Here are the possible designs that I see:

For `ReqBody`:

1. Pick the first one in the list.
  - pros:
    - easy implementation
    - no change to generated client functions
  - cons:
    - no flexibility to choose the used type
2. Use only one content-type supplied globally.
  - pros:
  - cons:
    - doesn't allow content-types by endpoint
3. Have additional proxy arguments to the generated functions (1 for body, 1 for result).
  - pros:
    - maximal flexibility
  - cons:
    - more arguments to generated functions
4. Do auto-matching somehow (Typeclasses for choices)
  - pros:
    - simpler function calls
  - cons:
    - much magic

Personally I prefer option 3. I think it should also be possible to implement 1 and 2 in terms of 3.

**Comment by mf59816** [comment]
After my second attempt at a draft for a pull request, I thought I'd wait for feedback from @jkarni or @alpmestan.  Also, I've had little time to work on this since then.

Your list looks helpful.  I will take a look and try to get a pull request done soon (this week-ish?).

**Comment by tvh** [comment]
I actually had a crack at 3, but could't get it to work. I then went and implemented 1: https://github.com/anchor/servant-client/tree/content-types. I based it off some generic changes (https://github.com/haskell-servant/servant-client/pull/10). It builds, but currently fails in the tests, because `decodeLenient` changed to `deocde`.

I think that approach is still fine. We get all the benefits of specifying the content type, so we can still have clients other than Haskell using different content-types. Having the extra power to choose the content type would be nice, but I couldn't think of a nice way to achieve that.

**Comment by mf59816** [comment]
+1 for option 1.

It seems quite restrictive to me, but we can always implement the other options later.

**Comment by jkarni** [comment]
@tvh I don't have that much against 1, since the semantic significance of ReqBody content-type list is still up for grabs. But out of curiosity, could you explain (4) a bit more?

**Comment by tvh** [comment]
@jkarni Table above holds for both `ReqBody` and the response types. For resposes we have an additional option 5 though:

- Return the result wrapped in `newtype WrappedResult ct a = WrappedResult a` or `type WrappedResult ct a = (Proxy ct, a)`.
  - pros:
    - additional flexibiltiy
  - cons:
    - pollution of result type

Option 4 is a variant of the approaches discussed above. You end up with typeclasses defining your choices. I think the most elegant design I saw so far doing this was

```.haskell
class ReqCT (layout, ctyps) where
    type ReqCTChoice (layout, ctyps)

class RespCT (layout, ctyps) where
    type RespCTChoice (layout, ctyps)
```

The main reason why I don't like that approach is, that it adds another layer of complexity to the system. As a user I would prefer not having to implement typeclasses to use servant.

**Comment by tvh** [comment]
@jkarni I am not sure what you mean by "the semantic significance of ReqBody content-type list is still up for grabs". To implement OAuth2 this has to be `application/x-www-form-urlencoded` for example.

**Comment by tvh** [comment]
I just put together an implementation for `application/x-www-form-urlencoded`. It currently uses `Network.URL` to decode and encode, but this leads to quite some marshalling.

https://gist.github.com/tvh/5e24f4818eb519127f0f

**Comment by alpmestan** [comment]
The vote about the naming convention happens here: https://github.com/haskell-servant/servant/issues/12

File: src/Servant/Server/Internal.hs [hunk 1]
```diff
@@ -261,17 +261,24 @@ instance HasServer Delete where
...
   route Proxy action request respond
     | pathIsEmpty request && requestMethod request == methodGet = do
         e <- runEitherT action
         respond . succeedWith $ case e of
-          Right output ->
-            responseLBS ok200 [("Content-Type", "application/json")] (encode output)
+          Right output -> do
+            let accH = fromMaybe "*/*" $ lookup hAccept $ requestHeaders request
+            case handleAcceptH (Proxy :: Proxy ctypes) (AcceptHeader accH) output of
+              Nothing -> responseLBS (mkStatus 406 "") [] ""
```
Comment (tvh) [inline src/Servant/Server/Internal.hs]: 406 Not Acceptable
